gpflow/kernels/convolutional.py

```diff
@@ -54,49 +61,111 @@ def __init__(
             np.ones(self.num_patches, dtype=default_float()) if weights is None else weights
         )
 
-    # @lru_cache() -- Can we do some kind of memoizing with TF2?
+    @check_shapes(
+        "X: [batch..., N, D]",
+        "return: [batch..., N, P, S]",
+    )
     def get_patches(self, X: TensorType) -> tf.Tensor:
         """
         Extracts patches from the images X. Patches are extracted separately for each of the colour
         channels.
 
-        :param X: (N x input_dim)
-        :return: Patches (N, num_patches, patch_shape)
+        :param X: Images.
+        :return: Patches.
         """
         # Roll the colour channel to the front, so it appears to
         # `tf.extract_image_patches()` as separate images. Then extract patches
         # and reshape to have the first axis the same as the number of images.
         # The separate patches will then be in the second axis.
-        num_data = tf.shape(X)[0]
-        castX = tf.transpose(tf.reshape(X, [num_data, -1, self.colour_channels]), [0, 2, 1])
-        patches = tf.image.extract_patches(
-            tf.reshape(castX, [-1, self.image_shape[0], self.image_shape[1], 1], name="rX"),
-            [1, self.patch_shape[0], self.patch_shape[1], 1],
-            [1, 1, 1, 1],
-            [1, 1, 1, 1],
-            "VALID",
+        batch = tf.shape(X)[:-2]
+        N = tf.shape(X)[-2]
+        D = tf.shape(X)[-1]
+        flat_batch = tf.reduce_prod(batch)
+        num_data = tf.reduce_prod(batch) * N
+        X = cs(
+            tf.transpose(tf.reshape(X, [num_data, -1, self.colour_channels]), [0, 2, 1]),
+            "[C, num_data, W_x_H]",
```

Comment:
Please check, I'm not a 100% sure
```suggestion
            "[num_data, C, W_x_H]",
```